File: scala-3/grocery-todo-list/src/main/scala/com/rbleggi/grocerytodolist/CommandInvoker.scala
package com.rbleggi.grocerytodolist

import scala.collection.mutable

object CommandInvoker {
  private val history = mutable.Stack[Command]()
  private val redoStack = mutable.Stack[Command]()

  def executeCommand(command: Command, items: List[GroceryItem]): List[GroceryItem] = {
    val result = command.execute(items)
    history.push(command)
    redoStack.clear()
    result
  }

  def undo(items: List[GroceryItem]): List[GroceryItem] = {
    if (history.nonEmpty) {
      val command = history.pop()
      val result = command.undo(items)
      redoStack.push(command)
      println("Undo successful")
      result
    } else {
      println("Nothing to undo")
      items
    }
  }

  def redo(items: List[GroceryItem]): List[GroceryItem] = {
    if (redoStack.nonEmpty) {
      val command = redoStack.pop()
      val result = command.execute(items)
      history.push(command)
      println("Redo successful")
      result
    } else {
      println("Nothing to redo")
      items
    }
  }
}

File: java/grocery-todo-list/src/main/java/com/rbleggi/grocerytodolist/CommandInvoker.java
package com.rbleggi.grocerytodolist;

import java.util.List;
import java.util.Stack;

public class CommandInvoker {
    private final Stack<Command> history = new Stack<>();
    private final Stack<Command> redoStack = new Stack<>();

    public List<GroceryItem> executeCommand(Command command, List<GroceryItem> items) {
        List<GroceryItem> result = command.execute(items);
        history.push(command);
        redoStack.clear();
        return result;
    }

    public List<GroceryItem> undo(List<GroceryItem> items) {
        if (!history.isEmpty()) {
            Command command = history.pop();
            List<GroceryItem> result = command.undo(items);
            redoStack.push(command);
            System.out.println("Undo successful");
            return result;
        } else {
            System.out.println("Nothing to undo");
            return items;
        }
    }

    public List<GroceryItem> redo(List<GroceryItem> items) {
        if (!redoStack.isEmpty()) {
            Command command = redoStack.pop();
            List<GroceryItem> result = command.execute(items);
            history.push(command);
            System.out.println("Redo successful");
            return result;
        } else {
            System.out.println("Nothing to redo");
            return items;
        }
    }
}

